internal/services/oidc.go

```diff
@@ -0,0 +1,250 @@
+package services
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/ydataai/authentication-service/internal/clients"
+	"github.com/ydataai/authentication-service/internal/models"
+	"github.com/ydataai/go-core/pkg/common/logging"
+	"golang.org/x/oauth2"
+)
+
+// OIDCService defines the oidc server struct.
+type OIDCService struct {
+	client         *clients.OIDCClient
+	sessionStorage *models.SessionStorage
+	logger         logging.Logger
+}
+
+// NewOIDCService creates a new OIDC Service.
+func NewOIDCService(logger logging.Logger,
+	client *clients.OIDCClient,
+	sessionStorage *models.SessionStorage) *OIDCService {
+	return &OIDCService{
+		client:         client,
+		sessionStorage: sessionStorage,
+		logger:         logger,
+	}
+}
+
+// GetReadyzFunc make sure if oidc provider is ready.
+func (osvc *OIDCService) GetReadyzFunc() func() bool {
+	return osvc.client.ReadyzFunc
+}
+
+// GetOIDCProviderURL gets OIDC provider URL from the OAuth2 configuration.
+func (osvc *OIDCService) GetOIDCProviderURL(w http.ResponseWriter, r *http.Request) string {
+	// Sets a temporary session with state and nonce to validate
+	// when there is callback from the OIDC Provider.
+	session := osvc.sessionStorage.CreateSession()
```

Comment:
the logic of the creation should be in here and the storage handling the access to it, not creating.